localstack/http/proxy.py

```diff
@@ -0,0 +1,151 @@
+    information.
+    """
+    return Proxy(forward_base_url=forward_base_url).forward(
+        request, forward_path=forward_path, headers=headers
+    )
+
+
+class Proxy:
+    def __init__(self, forward_base_url: str, client: HttpClient = None):
+        """
+        Creates a new HTTP Proxy which can be used to forward incoming requests according to the configuration.
+
+        :param forward_base_url: the base url (backend) to forward the requests to.
+        :param client: the HTTP Client used to make the requests
+        """
+        self.forward_base_url = forward_base_url
+        self.client = client or SimpleRequestsClient()
+
+    def forward(
+        self,
+        request: Request,
+        forward_path: str = None,
+        headers: Union[Headers, Mapping[str, str]] = None,
+    ):
+        """
+        Uses the client to forward the given request according to the proxy's configuration.
+
+        :param request: the base request to forward (with the original URL and path data)
+        :param forward_path: the path to forward the request to. if set, the original path will be replaced completely,
+            otherwise the original path will be used
+        :param headers:
+        :return:
+        """
+        headers = Headers(headers) if headers else Headers()
+
+        if client_ip := request.remote_addr:
+            if xff := request.headers.get("X-Forwarded-For"):
+                headers["X-Forwarded-For"] = f"{xff}, {client_ip}"
+            else:
+                headers["X-Forwarded-For"] = f"{client_ip}"
+
+        if forward_path is None:
+            forward_path = request.path
+        if forward_path:
+            forward_path = "/" + forward_path.lstrip("/")
+
+        proxy_request = _copy_request(request, self.forward_base_url, forward_path, headers)
+        return self.client.request(proxy_request)
+
+
+class ForwardingHandler:
+    """
+    A dispatcher Handler which can be used in a ``Router[Handler]``  that proxies incoming requests according to the
+    configuration.
+    """
+
+    def __init__(self, proxy: Proxy, forward_path_parameter: bool = True):
+        """
+        Creates a new Handler.
+
+        :param forward_path_parameter: if set to true, the forwarder will use the `path` argument from the keyword
+            arguments passed to the endpoint to forward, s.t. you can add rules like ("/foo/<path:path>") that will
+            implicitly strip `/foo` from the request.
+        """
+        self.proxy = proxy
+        self.forward_path_parameter = forward_path_parameter
+
+    def __call__(self, request: Request, **kwargs) -> Response:
+        if self.forward_path_parameter:
+            forward_path = kwargs.pop("path", "")
+        else:
+            forward_path = request.path
+
+        return self.proxy.forward(request, forward_path=forward_path)
+
+
+def path_forwarder(forward_base_url: str) -> ForwardingHandler:
```

Comment:
nit: I think the terminology is a bit confusing with the `forward` function.
This one basically creates a handler for you (so it's some kind of ForwardingHandler factory), while the `forward` function directly forwards a request to the proxy target.
But I guess that's quite subjective. The docs and the signature is very clear about what it does.

Reply:
fair point! maybe we should just remove this one here, rename `ForwardingHandler` to `ProxyHandler`, and make the constructor easier to use.